infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/ImagePullSecretProvisioner.java

```diff
@@ -0,0 +1,139 @@
+import io.fabric8.kubernetes.api.model.Secret;
+import io.fabric8.kubernetes.api.model.SecretBuilder;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.util.Base64;
+import java.util.List;
+import javax.inject.Inject;
+import org.eclipse.che.api.core.model.workspace.runtime.RuntimeIdentity;
+import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.infrastructure.docker.client.UserSpecificDockerRegistryCredentialsProvider;
+import org.eclipse.che.infrastructure.docker.client.dto.AuthConfigs;
+import org.eclipse.che.workspace.infrastructure.kubernetes.KubernetesClientFactory;
+import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
+import org.eclipse.che.workspace.infrastructure.kubernetes.namespace.KubernetesNamespaceFactory;
+
+/** @author David Festal */
+public class ImagePullSecretProvisioner implements ConfigurationProvisioner<KubernetesEnvironment> {
+
+  static final String SECRET_NAME = "workspace-private-registries";
+
+  private final UserSpecificDockerRegistryCredentialsProvider
+      userSpecificDockerRegistryCredentialsProvider;
+  private final KubernetesClientFactory clientFactory;
+  private final KubernetesNamespaceFactory namespaceFactory;
+
+  @Inject
+  public ImagePullSecretProvisioner(
+      UserSpecificDockerRegistryCredentialsProvider userSpecificDockerRegistryCredentialsProvider,
+      KubernetesClientFactory clientFactory,
+      KubernetesNamespaceFactory namespaceFactory) {
+    this.userSpecificDockerRegistryCredentialsProvider =
+        userSpecificDockerRegistryCredentialsProvider;
+    this.clientFactory = clientFactory;
+    this.namespaceFactory = namespaceFactory;
+  }
+
+  @Override
+  public void provision(KubernetesEnvironment k8sEnv, RuntimeIdentity identity)
+      throws InfrastructureException {
+
+    AuthConfigs authConfigs = userSpecificDockerRegistryCredentialsProvider.getCredentials();
+    if (authConfigs.getConfigs().isEmpty()) {
+      return;
+    }
+
+    Gson gson = new Gson();
+
+    Base64.Encoder encoder = Base64.getEncoder();
+
+    String config;
+    try (StringWriter strWriter = new StringWriter();
+        JsonWriter jsonWriter = gson.newJsonWriter(strWriter)) {
+      jsonWriter.beginObject();
+
+      authConfigs
+          .getConfigs()
+          .forEach(
+              (name, authConfig) -> {
+                try {
+                  if (!name.startsWith("https://")) {
+                    name = "https://" + name;
+                  }
+                  jsonWriter.name(name);
+                  jsonWriter.beginObject();
+                  jsonWriter.name("username");
+                  jsonWriter.value(authConfig.getUsername());
+                  jsonWriter.name("password");
+                  jsonWriter.value(authConfig.getPassword());
+                  jsonWriter.name("email");
+                  jsonWriter.value("email@email");
+                  String auth =
```

Comment:
Can you point me to a doc that describes how this object should look like - I don't understand by looking at the code. This also indicates that this code is not clear and might require additional code simplifications or comments to ease support of this code.